modules/actions/pool/upsert-pools.ts

```diff
@@ -0,0 +1,140 @@
+    }
+
+    // Get the token prices needed for calculating token balances and total liquidity
+    const dbPrices = await prisma.prismaTokenPrice.findMany({
+        where: {
+            tokenAddress: { in: allTokens.map((token) => token.address) },
+            chain: chain,
+        },
+    });
+
+    // Build helper maps for token prices and decimals
+    const decimals = Object.fromEntries(allTokens.map((token) => [token.address, token.decimals]));
+    const prices = Object.fromEntries(dbPrices.map((price) => [price.tokenAddress, price.price]));
+
+    // Get the data for the tables about pools
+    const dbPools: CompletePoolDbEntry[] = subgraphPools.map((poolData) => {
+        const onchainPoolData = onchainData[poolData.id];
+        const onchainTokensData = Object.fromEntries(onchainPoolData.tokens.map((token) => [token.address, token]));
+
+        return {
+            pool: poolTransformer(poolData, chain),
+            poolDynamicData: {
+                id: poolData.id,
+                poolId: poolData.id,
+                chain: chain,
+                totalShares: String(onchainPoolData.totalSupply),
+                totalSharesNum: Number(formatUnits(onchainPoolData.totalSupply, 18)),
+                blockNumber: Number(blockNumber),
+                swapFee: String(onchainPoolData.swapFee ?? '0'),
+                swapEnabled: true,
+                totalLiquidity: 0,
+            },
+            poolToken: poolTokensTransformer(poolData, chain),
+            poolTokenDynamicData: poolTokensDynamicDataTransformer(
+                poolData,
+                onchainTokensData,
+                decimals,
+                prices,
+                chain,
+            ),
+            poolExpandedTokens: poolData.tokens.map(({ address, nestedPool }) => ({
+                tokenAddress: address,
+                poolId: poolData.id,
+                chain: chain,
+                nestedPoolId: nestedPool?.id,
+            })),
+        };
+    });
+
+    // Upsert pools to the database
+    for (const { pool, poolToken, poolDynamicData, poolTokenDynamicData, poolExpandedTokens } of dbPools) {
+        try {
+            await prisma.prismaPool.upsert({
+                where: { id_chain: { id: pool.id, chain: pool.chain } },
+                create: pool,
+                update: pool,
+            });
+
+            await prisma.prismaPoolDynamicData.upsert({
+                where: { poolId_chain: { poolId: pool.id, chain: pool.chain } },
+                create: poolDynamicData,
+                update: poolDynamicData,
+            });
+
+            // First nullify the pool tokens and then insert them again
+            await prisma.prismaPoolToken.deleteMany({ where: { poolId: pool.id } });
+            await prisma.prismaPoolTokenDynamicData.deleteMany({ where: { poolTokenId: { startsWith: pool.id } } });
+            await prisma.prismaPoolExpandedTokens.deleteMany({ where: { poolId: pool.id } });
```

Comment:
is that needed? Wonder if we could have unexpected side effects if `upsertPools` is called regularly. Then pools end up having no tokens quite often.

Reply:
Ok, I see the problem. There is no `upsertMany` that works properly. hmm, need to add a todo here

Reply:
we can wrap it in a transaction, so there won't be intermittent "empty" state